### PR TITLE
Implemented merge sort 

### PR DESCRIPTION
**Pull request by Aditya1001001**
Sorting algorithms implemented (#8 ):
- [ ] quicksort
- [x] mergesort
- [ ] heapsort

### CODE REVIEW

File: pydsx/singly_linked_list/linked_list.py [hunk 1]
```diff
@@ -172,6 +175,148 @@ def __setitem__(self, index, new_value):
...
+            if self._head.next == None:
+                return self 
+            if key == "merge":
+                return self.merge_sort()
+            elif key == "quick":
+                pass
+            elif key == "heap":
+                pass
+            else:
+                raise ValueError("Unknown sorting method specified")
```
Comment (matheusfrancisco) [inline pydsx/singly_linked_list/linked_list.py]: I was looking at maybe we can try to apply a (DRY, don't reapt yourself reference for DRY (refactoring book))

May you can extract to a  method?
something like this

```python
self._sort_strategy(reverse, key)
#and self._merge_sort(...) choose if reversed or not
def _sort_strategy(self, reverse, key):
    if self._head == None:
        raise IndexError("The linked list is empty")       
    if self._head.next == None:
        return self 
    if key == "merge":
        return self.merge_sort(reverse)
    elif key == "quick":
        pass
    elif key == "heap":
        pass
    else:
       raise ValueError("Unknown sorting method specified")
       
```
Does that make sense to you?

I'll keep reviewing, thanks !!
Congratulations on the code

File: tests/test_linked_list.py [hunk 2]
```diff
@@ -125,3 +125,50 @@ def test_append_with_tail_flag():
...
+    assert l._head.next.value == 2
+    assert l._head.next.next.value == 3
+    assert l._head.next.next.next.value == 4
+    assert l._head.next.next.next.next.value == 5
+    assert l._head.next.next.next.next.next.value == 6
+    assert l._head.next.next.next.next.next.next.value == 7
+    assert l._head.next.next.next.next.next.next.next.value == 8
+    assert l._head.next.next.next.next.next.next.next.next.value == 9
+    assert l._head.next.next.next.next.next.next.next.next.next == None
+
```
Comment (matheusfrancisco) [inline tests/test_linked_list.py]: You can write less code in this test by doing so..
```python
    list_to_compare = [1, 2, 3, 4, 5, 6, 7, 8, 9]

    #has_same_size = len(list_to_compare) == len(l)
    has_equal_values = all(
        map(lambda x: x[0] == x[1], zip(l, list_to_compare))
    )
    assert has_equal_values is True
    #assert has_same_size is True
```

File: pydsx/singly_linked_list/linked_list.py (hunk 1 above)
Comment (Aditya1001001) [reply pydsx/singly_linked_list/linked_list.py]: That does make sense, I changed it up while trying to fix a bug in the  ```merge_sort``` method, I'll put it back to how it was.

File: tests/test_linked_list.py (hunk 2 above)
Comment (Aditya1001001) [reply tests/test_linked_list.py]: Done!

**Review by matheusfrancisco — APPROVED** [review]

**Comment by matheusfrancisco** [comment]
thanks for your pull request!

**Pull request merged**